File: backend/app/services/admin_role_requests.py
from __future__ import annotations

import uuid
from datetime import datetime, timezone

from sqlalchemy import Select, desc, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.models.admin_role_request import AdminRoleRequest, AdminRoleRequestStatus
from app.models.user import User, UserRole


class AdminRoleRequestConflictError(Exception):
    pass


class AdminRoleRequestNotFoundError(Exception):
    pass


class AdminRoleRequestInvalidStateError(Exception):
    pass


def _list_requests_stmt() -> Select[tuple[AdminRoleRequest, User]]:
    return (
        select(AdminRoleRequest, User)
        .join(User, User.id == AdminRoleRequest.user_id)
        .order_by(desc(AdminRoleRequest.created_at))
    )


async def create_admin_role_request(session: AsyncSession, user: User) -> AdminRoleRequest:
    if user.role == UserRole.admin:
        raise AdminRoleRequestConflictError("У вас уже есть права администратора")

    existing_pending = await session.execute(
        select(AdminRoleRequest).where(
            AdminRoleRequest.user_id == user.id,
            AdminRoleRequest.status == AdminRoleRequestStatus.pending,
        )
    )
    if existing_pending.scalar_one_or_none() is not None:
        raise AdminRoleRequestConflictError("Заявка уже отправлена и ожидает рассмотрения")

    req = AdminRoleRequest(user_id=user.id, status=AdminRoleRequestStatus.pending)
    session.add(req)
    await session.commit()
    await session.refresh(req)
    return req


async def get_my_latest_request(session: AsyncSession, user_id: uuid.UUID) -> tuple[AdminRoleRequest, User] | None:
    result = await session.execute(
        _list_requests_stmt().where(AdminRoleRequest.user_id == user_id).limit(1)
    )
    row = result.first()
    if row is None:
        return None
    return row[0], row[1]


async def list_admin_role_requests(
    session: AsyncSession,
    *,
    status: AdminRoleRequestStatus | None = None,
) -> list[tuple[AdminRoleRequest, User]]:
    stmt = _list_requests_stmt()
    if status is not None:
        stmt = stmt.where(AdminRoleRequest.status == status)
    result = await session.execute(stmt)
    return list(result.all())


async def review_admin_role_request(
    session: AsyncSession,
    *,
    request_id: uuid.UUID,
    reviewer: User,
    approve: bool,
    review_comment: str | None = None,
) -> tuple[AdminRoleRequest, User]:
    result = await session.execute(
        select(AdminRoleRequest, User)
        .join(User, User.id == AdminRoleRequest.user_id)
        .where(AdminRoleRequest.id == request_id)
    )
    row = result.first()
    if row is None:
        raise AdminRoleRequestNotFoundError("Заявка не найдена")
    req, target_user = row

    if req.status != AdminRoleRequestStatus.pending:
        raise AdminRoleRequestInvalidStateError("Заявка уже рассмотрена")

    req.status = AdminRoleRequestStatus.approved if approve else AdminRoleRequestStatus.rejected
    req.reviewed_by = reviewer.id
    req.review_comment = review_comment
    req.reviewed_at = datetime.now(timezone.utc)

    if approve:
        target_user.role = UserRole.admin

    await session.commit()
    await session.refresh(req)
    await session.refresh(target_user)
    return req, target_user

File: frontend/src/api/adminRoleRequests.ts
import { apiFetch } from "./client";

export type AdminRoleRequestStatus = "pending" | "approved" | "rejected";

export type AdminRoleRequest = {
  id: string;
  user_id: string;
  user_email: string;
  status: AdminRoleRequestStatus;
  review_comment: string | null;
  reviewed_by: string | null;
  reviewed_at: string | null;
  created_at: string;
  updated_at: string;
};

export function adminRoleRequestStatusLabel(status: AdminRoleRequestStatus | null | undefined): string {
  if (status === "pending") return "на рассмотрении";
  if (status === "approved") return "одобрена";
  if (status === "rejected") return "отклонена";
  return "не подана";
}

export async function createAdminRoleRequest(): Promise<void> {
  await apiFetch("/api/v1/admin-role-requests/", {
    method: "POST",
  });
}

export async function getMyAdminRoleRequest(): Promise<AdminRoleRequest | null> {
  return apiFetch<AdminRoleRequest | null>("/api/v1/admin-role-requests/me");
}

export async function listAdminRoleRequests(status?: AdminRoleRequestStatus): Promise<AdminRoleRequest[]> {
  const q = status ? `?status=${encodeURIComponent(status)}` : "";
  return apiFetch<AdminRoleRequest[]>(`/api/v1/admin-role-requests/${q}`);
}

export async function approveAdminRoleRequest(id: string, reviewComment?: string): Promise<AdminRoleRequest> {
  return apiFetch<AdminRoleRequest>(`/api/v1/admin-role-requests/${id}/approve`, {
    method: "POST",
    body: JSON.stringify({ review_comment: reviewComment ?? null }),
  });
}

export async function rejectAdminRoleRequest(id: string, reviewComment?: string): Promise<AdminRoleRequest> {
  return apiFetch<AdminRoleRequest>(`/api/v1/admin-role-requests/${id}/reject`, {
    method: "POST",
    body: JSON.stringify({ review_comment: reviewComment ?? null }),
  });
}
